### PR TITLE
Rewrite the root docs

### PR DESCRIPTION
**Pull request by sgrif**
This is inspired by the module docs for `std`. The goal of these docs is
to help you find the most important modules, and generally learn where
to find things.

### CODE REVIEW

File: diesel/src/lib.rs [hunk 1]
```diff
@@ -1,16 +1,95 @@
...
+//! # Where to find things
 //!
-//! * **Diesel**: core libary, documented here
-//! * [**Diesel CLI**](https://github.com/diesel-rs/diesel/tree/master/diesel_cli): command line tool that aids in managing your database schema
+//! ## Declaring your schema
 //!
+//! Diesel ensures that your queries are valid at compile time.
+//! This means it needs to know what your schema is.
+//! This is done using [the `table!` macro][`table!`].
+//! This macro is usually generated by `diesel print-schema` or [`infer_schema!`]
```
Comment (notryanb) [inline diesel/src/lib.rs]: Lots of "this". Maybe reword this block to something similar to...
> Diesel ensures that your queries are valid at compile time by knowing about your schema. This is done using  [the `table!` macro][`table!`], which is usually generated by `diesel print-schema` or [`infer_schema!`]

?

Comment (killercup) [reply diesel/src/lib.rs]: Yeah. I want to keep the "it needs to know about" aspect, though.

> So Diesel is able to validate your queries at compile-time, it requires you to specify your database's schema in your code, which you can do with the `table!` macro. `diesel print-schema` or `infer_schema!` can be used to automatically generate these macro calls (by connecting to your database and parsing its schema).

File: diesel/src/lib.rs [hunk 2]
```diff
@@ -1,16 +1,95 @@
...
+//!
+//! [`table!`]: macro.table.html
+//! [`infer_schema!`]: FIXME: Host this crate on docs.diesel.rs because docs.rs won't render docs for this crate
```
Comment (killercup) [inline diesel/src/lib.rs]: (This comment is just to make this **FIXME** show up in the comments sections so it gets fixed before merging)

File: diesel/src/lib.rs [hunk 3]
```diff
@@ -1,16 +1,95 @@
...
+//!
+//! ## Getting started
+//!
+//! Queries usually start from either a table, or a function like [`update`].
+//! Those functions can be found [here](#functions).
+//!
+//! Diesel provides a [`prelude` module](prelude),
+//! which exports most of the most important traits and types.
```
Comment (killercup) [inline diesel/src/lib.rs]: Maybe `s/most of the most important/most of the typically used`?

File: diesel/src/lib.rs [hunk 4]
```diff
@@ -1,16 +1,95 @@
...
+//! We are conservative about what goes in this module,
+//! and avoid anything which has a generic name.
+//! Files which use Diesel are expected to have `use diesel::prelude::*;`.
```
Comment (killercup) [inline diesel/src/lib.rs]: > are expected to

What expects this? Our derives? Macros? Maybe add this as a reason

File: diesel/src/lib.rs [hunk 5]
```diff
@@ -1,16 +1,95 @@
...
+//! [`update`]: fn.update.html
+//!
+//! ## Constructing a query
+//!
+//! Diesel breaks things into three categories:
+//!
+//! - "Query builder methods" are things that map to portions of a whole query
+//!   (such as `ORDER` and `WHERE`). These methods usually have the same name
+//!   as the SQL they map to, except for `WHERE` which is called `filter` in Diesel.
+//!   These methods live in [the `query_dsl` module](query_dsl).
```
Comment (killercup) [inline diesel/src/lib.rs]: No semantics linefeeds in the paragraph?

File: diesel/src/lib.rs [hunk 6]
```diff
@@ -1,16 +1,95 @@
...
+//! which exports most of the most important traits and types.
+//! We are conservative about what goes in this module,
+//! and avoid anything which has a generic name.
+//! Files which use Diesel are expected to have `use diesel::prelude::*;`.
+//!
+//! [`update`]: fn.update.html
+//!
+//! ## Constructing a query
+//!
+//! Diesel breaks things into three categories:
```
Comment (killercup) [inline diesel/src/lib.rs]: These "things" are… what exactly? "Rust items"? "Operations"? "Functions"? "Methods"?

Maybe something like

> The tools the query builder gives you can be put into these three categories:

File: diesel/src/lib.rs [hunk 7]
```diff
@@ -1,16 +1,95 @@
...
+//!
+//! - "Query builder methods" are things that map to portions of a whole query
+//!   (such as `ORDER` and `WHERE`). These methods usually have the same name
+//!   as the SQL they map to, except for `WHERE` which is called `filter` in Diesel.
```
Comment (killercup) [inline diesel/src/lib.rs]: > (to not conflict with the Rust keyword)

File: diesel/src/lib.rs [hunk 8]
```diff
@@ -1,16 +1,95 @@
...
+//!   These methods live in [the `query_dsl` module](query_dsl).
+//! - "Expression methods" are things you would call on columns
+//!   or other individual values.
+//!   These methods live in [the `expression_methods` module](expression_methods)
+//!   You can often find these by thinking "what would this be called"
+//!   if it were a method
+//!   and typing that into the search bar
+//!   (e.g. `LIKE` is called `like` in Diesel).
+//!   Most operators are named based on the Rust function which maps to that
+//!   operator in `std::ops`
```
Comment (killercup) [inline diesel/src/lib.rs]: Not sure how many people are familiar with the ops module, most probably only use the operators themselves. So, maybe link to <https://doc.rust-lang.org/1.22.1/std/ops/index.html>?

File: diesel/src/lib.rs [hunk 9]
```diff
@@ -1,16 +1,95 @@
...
+//!
+//! Diesel breaks things into three categories:
+//!
+//! - "Query builder methods" are things that map to portions of a whole query
+//!   (such as `ORDER` and `WHERE`). These methods usually have the same name
+//!   as the SQL they map to, except for `WHERE` which is called `filter` in Diesel.
+//!   These methods live in [the `query_dsl` module](query_dsl).
+//! - "Expression methods" are things you would call on columns
+//!   or other individual values.
+//!   These methods live in [the `expression_methods` module](expression_methods)
```
Comment (killercup) [inline diesel/src/lib.rs]: This (and query_dsl) are weird relative links. I assume it's supposed to lead to `./expression_methods/index.html`?

File: diesel/src/lib.rs [hunk 10]
```diff
@@ -1,16 +1,95 @@
...
+//! [`sql_function!`]: macro.sql_function.html
+//!
+//! ## Serializing and Deserializing
+//!
+//! Types which represent the result of a SQL query implement
+//! a trait called [`Queryable`].
+//!
+//! Diesel maps "Rust types" (e.g. `i32`) to and from "SQL types"
+//! (e.g. [`diesel::types::Integer`]).
+//! You can find all the types supported by Diesel in [the `types` module](types).
```
Comment (killercup) [inline diesel/src/lib.rs]: Maybe also mention backend specific types?

**Comment by sgrif** [comment]
> I just hope they find it ;)

This is the most prominent place I can put it, so they better. >_>

File: diesel/src/lib.rs (hunk 4 above)
Comment (sgrif) [reply diesel/src/lib.rs]: It's just convention. All our docs are written assuming that line. All our macros and derives should work without it.

File: diesel/src/lib.rs (hunk 9 above)
Comment (sgrif) [reply diesel/src/lib.rs]: Yes. Static file servers typically render `index.html` if a directory is requested.

File: diesel/src/lib.rs (hunk 10 above)
Comment (sgrif) [reply diesel/src/lib.rs]: The docs for the `types` module shows links to backend specific types.

File: diesel/src/lib.rs (hunk 9 above)
Comment (killercup) [reply diesel/src/lib.rs]: Yeah, it just looked like a markdown ref link at first :)

**Comment by sgrif** [comment]
I was going to wait for the build but ok :P

**Comment by killercup** [comment]
Is there anything in here that is tested by CI? Did you add a link checker without me noticing? :D

**Comment by sgrif** [comment]
No, I just generally wait all the time now after that time that I broke CI like 3 times in a day with PRs that "definitely aren't doing anything that could break CI"

**Comment by killercup** [comment]
Pfff and you expect to learn from _your_ failures? I hardly learn from my own!
